File: problem69.py
import math

def hcf(num1,num2):
    return math.gcd(num1,num2)


answer = 0
for i in range(1_000_000):
    count = 0
    if i % 10000 == 0:
        print(f"{i}/{1_000_000} ({i/1_000_000*100}%)")
    for j in range(1,i+1):
        if hcf(i,j) == 1:
            count +=1
    try:
        answer = max(answer,i/count)
    except:
        pass
print("answer: ",answer)

File: problem69.c
#include <math.h>
#include <stdio.h>
#define LIMIT 1000000

int gcd(int a, int b) {
    while (b != 0) {
        int t = b;
        b = a % b;
        a = t;
    }
    return a;
}


int main() {
    int *phi = malloc((LIMIT + 1) * sizeof(int));

    for (int i = 0; i <= LIMIT; i++)
        phi[i] = i;  // Initialize φ(n) = n

    // Euler's Totient sieve
    for (int i = 2; i <= LIMIT; i++) {
        if (phi[i] == i) { // i is prime
            for (int j = i; j <= LIMIT; j += i) {
                phi[j] -= phi[j] / i;
            }
        }
    }

    // Find max φ(n)
    int answer = 0;
    int number = 0;
    for (int i = 1; i <= LIMIT; i++) {
        if (phi[i] > answer) {
            answer = phi[i];
            number = i;
        }

        if (i % 100000 == 0)
            printf("%d/1000000 processed\n", i);
    }

    printf("Max φ(n) = %d (at n = %d)\n", answer, number);

    free(phi);
    getchar();
    return 0;
}